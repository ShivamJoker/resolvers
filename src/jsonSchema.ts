import { appendErrors, Resolver, transformToNestObject } from 'react-hook-form';
//import * as Ajv from 'ajv';
import Ajv, { ErrorObject, Options } from 'ajv';
import type { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';
import {
  FieldError,
  FieldErrors,
  ResolverError,
  ResolverSuccess,
} from 'react-hook-form/dist/types/form';

export type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;

const parseErrorSchema = <TFieldValues extends Record<string, any>>(
  validationError: Array<ErrorObject> | null | undefined,
  validateAllFieldCriteria: boolean,
): FieldErrors<TFieldValues> =>
  Array.isArray(validationError)
    ? validationError.reduce(
        (
          previous: FieldErrors<TFieldValues>,
          { dataPath, keyword, message },
        ) => {
          const path = dataPath.replace(/\//g, '.').replace(/^\./, '');
          return {
            ...previous,
            ...(path
              ? previous[path] && validateAllFieldCriteria
                ? {
                    [path]: appendErrors(
                      path,
                      validateAllFieldCriteria,
                      previous as Record<string, any>,
                      keyword,
                      message,
                    ),
                  }
                : {
                    [path]:
                      previous[path] ||
                      ({
                        type: keyword,
                        message,
                        ...(validateAllFieldCriteria
                          ? {
                              types: { [keyword]: message || true },
                            }
                          : {}),
                      } as FieldError),
                  }
              : {}),
          };
        },
        {},
      )
    : {};

export interface JsonSchemaOptions {
  ajvOptions?: Omit<
    Options,
    'allErrors' | 'async' | 'validateSchema' | 'transpile'
  >;
}

export const jsonSchemaResolver = <TFieldValues extends Record<string, any>>(
  validationSchema: JSONSchema,
  options: JsonSchemaOptions = {},
): Resolver<TFieldValues> => {
  if (!validationSchema || typeof validationSchema !== 'object') {
    throw new Error('Invalid AJV schema or validation function');
  }

  const ajv = new Ajv({
    ...options.ajvOptions,
    async: false,
    allErrors: true,
    validateSchema: true,
    transpile: undefined,
  });
  const validate = ajv.compile(validationSchema);

  return async (data, _, validateAllFieldCriteria = false) => {
    const valid = validate(data) as boolean;
    const errors = validate.errors;

    return valid
      ? ({
          values: data,
          errors: {},
        } as ResolverSuccess<TFieldValues>)
      : ({
          values: {},
          errors: transformToNestObject(
            parseErrorSchema<TFieldValues>(errors, validateAllFieldCriteria),
          ),
        } as ResolverError<TFieldValues>);
  };
};
