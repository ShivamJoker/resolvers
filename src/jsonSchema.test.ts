import { JSONSchema, jsonSchemaResolver } from './jsonSchema';

const syncSchema: JSONSchema = {
  type: 'object',
  properties: {
    name: {
      type: 'string',
      pattern: '[a-zA-Z]',
      minLength: 3,
    },
    age: {
      type: 'integer',
      minimum: 0,
    },
    email: {
      type: 'string',
      format: 'email',
    },
    hobbies: {
      type: 'array',
      minItems: 1,
      items: {
        properties: {
          description: {
            type: 'string',
            minLength: 3,
          },
        },
        required: ['description'],
      },
    },
  },
  required: ['name', 'age', 'email', 'hobbies'],
};

const asyncSchema: JSONSchema = {
  ...syncSchema,
  $async: true,
};

describe('ajvResolver', () => {
  describe('sync', () => {
    it('should get values when there is no error', async () => {
      const data = {
        name: 'jimmy',
        age: 24,
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: 'tennis',
          },
        ],
      };
      expect(await jsonSchemaResolver(syncSchema)(data)).toEqual({
        values: data,
        errors: {},
      });
    });

    it('should get errors on invalid data', async () => {
      const data = {
        name: 'jimmy',
        age: '24',
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: 'tennis',
          },
          {
            description: 'ab',
          },
        ],
      };

      expect(await jsonSchemaResolver(syncSchema)(data)).toMatchSnapshot();
    });

    it('should throw when an invalid schema object is provided', async () => {
      expect(() => {
        jsonSchemaResolver({
          ...syncSchema,
          type: 'o' as any,
        });
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw when an invalid schema param is provided to ajv', async () => {
      expect(() => {
        jsonSchemaResolver(jsonSchemaResolver as any);
      }).toThrowErrorMatchingSnapshot();
    });
  });
  describe('async', () => {
    it('should get values when there is no error', async () => {
      const data = {
        name: 'jimmy',
        age: 24,
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: 'tennis',
          },
        ],
      };
      expect(await jsonSchemaResolver(asyncSchema)(data)).toEqual({
        values: data,
        errors: {},
      });
    });

    it('should get errors on invalid data', async () => {
      const data = {
        name: 'jimmy',
        age: '24',
        email: 'jimmy@jimmy.com',
        hobbies: [
          {
            description: 'tennis',
          },
          {
            description: 'ab',
          },
        ],
      };

      expect(
        await jsonSchemaResolver(asyncSchema, { ajvOptions: {} })(
          data,
          undefined,
          true,
        ),
      ).toMatchSnapshot();
    });

    it('should throw when an invalid schema object is provided', async () => {
      expect(() => {
        jsonSchemaResolver(
          {
            ...syncSchema,
            type: 'o' as any,
          },
          {},
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw when an invalid schema param is provided to ajv', async () => {
      expect(() => {
        jsonSchemaResolver(jsonSchemaResolver as any, {
          ajvOptions: { async: true },
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
