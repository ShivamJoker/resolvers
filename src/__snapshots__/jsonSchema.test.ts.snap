// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ajvResolver should get errors on invalid data 1`] = `
Object {
  "errors": Object {
    "age": Object {
      "message": "should be integer",
      "type": "type",
      "types": Object {
        "type": "should be integer",
      },
    },
    "hobbies": Array [
      undefined,
      Object {
        "description": Object {
          "message": "should NOT be shorter than 3 characters",
          "type": "minLength",
          "types": Object {
            "minLength": "should NOT be shorter than 3 characters",
          },
        },
      },
    ],
  },
  "values": Object {},
}
`;

exports[`ajvResolver should throw when an invalid schema object is provided 1`] = `"schema is invalid: data.type should be equal to one of the allowed values, data.type should be array, data.type should match some schema in anyOf"`;

exports[`ajvResolver should throw when an invalid schema param is provided to ajv 1`] = `"Invalid AJV schema or validation function"`;
